File: bindings/javascript/aal/math/matrix.js
const Aal = require("../gir.js").Aal
const AalWrapper = require("../wrapper.js").AalWrapper

exports.Matrix = class Matrix extends AalWrapper {
    get size() {
        return [this._object.getRowsNumber(), this._object.getColumnsNumber()]
    }

    get rows() {
        return this._object.getRowsNumber()
    }

    get columns() {
        return this._object.getColumnsNumber()
    }

    static sized(rows, columns) {
        const obj = Aal.Matrix.create(rows, columns)
        return new Matrix(obj)
    }

    static from_array(array) {
        const arr = []
        for (let i = 0; i < array.length; i++) {
            for (let j = 0; j < array[0].length; j++) {
                arr[i * array[0].length + j] = array[i][j]
            }
        }
        const res = new Matrix()
        res._object.setArray(arr, array.length)
        return res
    }

    get(i, j) {
        return this._object.getVal(i, j)
    }

    set(i, j, v) {
        this._object.setVal(i, j, v)
    }

    to_array() {
        const arr = new Array(this.rows)
        const ar = this._object.toArray()
        for (let i = 0; i < arr.length; i++) {
            arr[i] = new Array(this.columns)
            for (let j = 0; j < this.columns; j++) {
                arr[i][j] = ar[i * this.rows + j]
            }
        }
        return arr
    }

    copy() {
        result = new Matrix(this.rows, this.columns)
        result._object = this._object.copy()
        return result
    }

    get det() {
        return this._object.det()
    }

    get is_diagonal() {
        return this._object.isDiagonal()
    }
}

File: bindings/python/aal/math/matrix.py
from typing import Tuple
import numpy as np
from aal.gir import Aal
from aal.wrapper import AalWrapper

class Matrix(AalWrapper):
    SUPPORTED_TYPES = (Aal.Matrix, )

    def __init__(self, rows: int, columns: int):
        super().__init__()
        self._object = Aal.Matrix.create(rows, columns)

    @staticmethod
    def from_numpy(array: np.ndarray):
        result = Matrix(1, 1)
        result._object.set_array(array.reshape((array.size)).tolist(), array.shape[0])
        return result

    @property
    def size(self) -> Tuple[int, int]:
        return (self._object.get_rows_number(), self._object.get_columns_number())

    @property
    def rows(self) -> int:
        return self._object.get_rows_number()

    @property
    def columns(self) -> int:
        return self._object.get_columns_number()

    def __getitem__(self, c) -> float:
        x, y = c
        return self._object.get_val(x, y)

    def __setitem__(self, c, value):
        x, y = c
        self._object.set_val(x, y, value)        

    def to_numpy(self) -> np.ndarray:
        return np.asarray(self._object.to_array()).reshape((self.rows, self.columns))

    def copy(self):
        result = Matrix(self.rows, self.columns)
        result._object = self._object.copy()
        return result

    @property
    def det(self) -> float:
        return self._object.det()

    @property
    def is_diagonal(self) -> bool:
        return self._object.is_diagonal()

